pkg/server/sotw/v2/server.go

```diff
@@ -195,78 +169,13 @@ func (s *server) process(stream Stream, reqCh <-chan *discovery.DiscoveryRequest
 		case <-s.ctx.Done():
 			return nil
 		// config watcher can send the requested resources types in any order
-		case resp, more := <-values.endpoints:
-			if !more {
-				return status.Errorf(codes.Unavailable, "endpoints watch failed")
-			}
-			nonce, err := send(resp, resource.EndpointType)
-			if err != nil {
-				return err
-			}
-			values.endpointNonce = nonce
-
-		case resp, more := <-values.clusters:
-			if !more {
-				return status.Errorf(codes.Unavailable, "clusters watch failed")
-			}
-			nonce, err := send(resp, resource.ClusterType)
+		case resp := <-values.responses:
+			typeUrl := resp.GetRequest().TypeUrl
```

Comment:
Should we drain `values.responses` here? It might help with buffer overrun:
```
			for {
				typeUrl := resp.GetRequest().TypeUrl
				nonce, err := send(resp, typeUrl)
				if err != nil {
					return err
				}
				values.nonces[typeUrl] = nonce

				select {
				case resp = <-values.responses:
				default:
					break
				}
			}
```

Reply:
To be honest - I don't know. Do you really think it is worth doing and it will not add delays to processing of incoming requests from envoy?

Reply:
Don't know either. But thinking aloud, if server gets a lot of requests, it'll create a lot of watches that may be immediately pushed onto buffer. So even if go select is fair, it'll never clean up the buffer (each request generates produces one buffer item, each response removes one buffer item).

Reply:
"a lot of requests" in this particular case can be up to 8 :) our responses channel most of the time will be filled with  up to 8 resource types in ADS mode and 1 resource type in non-ADS mode, hence I think it all should work similarly to the way it worked before.
In any case - it is easy to add additional for loop as a separate PR.

Reply:
I have some questions about the buffered approach. At lyft we dont use ADS and use separate cds/eds/lds/rds channels. At scale EDS channel causes issues when the transport layer is overflowing and we sometimes drop messages. In this case we have devised mechanisms to ensure correctness by retry and dropping redundant updates for eds. In the PR, we are changing the model to use the same channel to use all xds. I feel at scale eds could starve other channels. Also, in case of a missed update, we have to device different mechanisms to ensure correctness for cds/rds/lds. This is an explicit ask for the control plane operators to implement. WDYT about this issue?

Reply:
If I understand it correctly - you won't use single channel in non-ADS case. this channel is created per stream, and each resource/resource type will create separate stream (hence separate channel).

Reply:
ah right..and we still have a way to specify a buffered channel with length 1.

Reply:
In this case - better chose 2. Cause it is possible to have a value written to channel before cancel() is called due to subsequent request from envoy coming.